mod-authentication/src/main/java/gg/solarmc/loader/authentication/AuthManager.java

```diff
@@ -0,0 +1,221 @@
+			However, we cannot permit multiple cracked users to have the same name ignoring case.
+			For premium users, it's OK because all we need to do is check if the UUID is premium.
+			For cracked users, it's not okay because the cracked users will have different UUIDs.
+			 */
+			return InitialLoginAttempt.deniedCaseSensitivityOfName();
+		}
+		VerifiablePassword existingPassword = new VerifiablePassword(
+				new PasswordHash(passwordRecord.get(AUTH_PASSWORDS.PASSWORD_HASH)),
+				new PasswordSalt(passwordRecord.get(AUTH_PASSWORDS.PASSWORD_SALT)),
+				new HashingInstructions(iterations, passwordRecord.get(AUTH_PASSWORDS.MEMORY))
+		);
+		return InitialLoginAttempt.needsPassword(existingPassword);
+	}
+
+	/**
+	 * Creates an automatic account. Called only for premium users
+	 * @param context the context
+	 * @param user the user
+	 */
+	private void insertAutomaticAccount(DSLContext context, UnauthenticatedUser user) {
+		context.insertInto(AUTH_PASSWORDS)
+				// iterations, being omitted from the insert, will be set to 0
+				.columns(AUTH_PASSWORDS.USERNAME, AUTH_PASSWORDS.PASSWORD_HASH, AUTH_PASSWORDS.PASSWORD_SALT)
+				.values(user.username(), passwordHasher.emptyHash(), saltGenerator.emptySalt())
+				.execute();
+		context.insertInto(USER_IDS)
+				.columns(USER_IDS.UUID)
+				.values(UUIDUtil.toByteArray(user.mcUuid()))
+				.onDuplicateKeyIgnore() // The premium user may have joined before with a different name
+				.execute();
+	}
+
+	/**
+	 * Creates the user's account. Should only be called when {@link InitialLoginAttempt.ResultType#NEEDS_ACCOUNT}
+	 * is encountered
+	 *
+	 * @param transaction the transaction
+	 * @param user the unauthenticated user
+	 * @param password the verifiable password
+	 * @return a create account result indicating success or failure
+	 */
+	public CreateAccountResult createAccount(Transaction transaction,
+											 UnauthenticatedUser user, VerifiablePassword password) {
+		DSLContext context = transaction.getProperty(DSLContext.class);
+		HashingInstructions instructions = password.instructions();
+
+		int updateCount = context
+				.insertInto(AUTH_PASSWORDS)
+				.columns(AUTH_PASSWORDS.USERNAME,
+						AUTH_PASSWORDS.ITERATIONS, AUTH_PASSWORDS.MEMORY,
+						AUTH_PASSWORDS.PASSWORD_HASH, AUTH_PASSWORDS.PASSWORD_SALT)
+				.values(user.username(),
+						(byte) instructions.iterations(), instructions.memory(),
+						password.passwordHash().hashUncloned(), password.passwordSalt().saltUncloned())
+				.onDuplicateKeyIgnore()
+				.execute();
+		if (updateCount == 0) {
+			return CreateAccountResult.CONFLICT;
+		}
+		context.insertInto(USER_IDS)
+				.columns(USER_IDS.UUID)
+				.values(UUIDUtil.toByteArray(user.mcUuid()))
+				.execute();
+		return CreateAccountResult.CREATED;
+	}
+
+	/**
+	 * Assuming the user has entered the password correctly, complete their login. <br>
+	 * <br>
+	 * Will migrate the user's previous data if they are now premium, but used to be cracked
```

Comment:
I like this! Although, i initially thought there would be a manual way of migrating accounts rather than automatically migrating if a premium user logged in, that way you could choose if you wanted to remain cracked or not.

Reply:
Yeah, I doubt anyone would want to remain cracked. There don't seem to be any advantages to it.